### PR TITLE
[WOR-73] Initial create MC workspace API and service implementation

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
In #1629 we threaded through the new `workspace_type` from the DB to the associated models. Now, we need to wire up and conditionally create workspaces in WSM, and potentially related Azure cloud contexts.

## This PR
* Creates a new `MultiCloudWorkspaceService` which knows how to provision an azure workspace. For now, this codepath is limited to creating workspaces with Azure cloud contexts, attempts to call with `Gcp` will result in an exception
* Gates access to that method by requiring a new config setting `multiCloudWorkspaces.enabled` be set to `true`, _and_ that the incoming request has the `cloudPlatform` field set
* Adds tests for the new service
* **Note**: This is a PR against the aforementioned #1629. 
## TODO
- [x] Tests

____

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala [hunk 1]
```diff
@@ -37,7 +37,7 @@ import org.broadinstitute.dsde.rawls.user.UserService
 import org.broadinstitute.dsde.rawls.util.ScalaConfig._
 import org.broadinstitute.dsde.rawls.util._
 import org.broadinstitute.dsde.rawls.webservice._
-import org.broadinstitute.dsde.rawls.workspace.WorkspaceService
+import org.broadinstitute.dsde.rawls.workspace.{MultiCloudWorkspaceService, WorkspaceService}
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:40]: Changes in this file are purely wiring up the new service during the boot process

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala [hunk 2]
```diff
@@ -25,7 +25,7 @@ import org.broadinstitute.dsde.rawls.openam.StandardUserInfoDirectives
 import org.broadinstitute.dsde.rawls.snapshot.SnapshotService
 import org.broadinstitute.dsde.rawls.status.StatusService
 import org.broadinstitute.dsde.rawls.user.UserService
-import org.broadinstitute.dsde.rawls.workspace.WorkspaceService
+import org.broadinstitute.dsde.rawls.workspace.{MultiCloudWorkspaceService, WorkspaceService}
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala:28]: More wire-up for consumption of the new service in the API

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 3]
```diff
@@ -140,6 +141,18 @@ object WorkspaceVersions {
...
 }
 
+
+case class MultiCloudWorkspaceRequest(namespace: String,
+                                      name: String,
+                                      attributes: AttributeMap,
+                                      authorizationDomain: Option[Set[ManagedGroupRef]] = Option(Set.empty),
+                                      cloudPlatform: WorkspaceCloudPlatform
+                                     ) extends Attributable {
+  def toWorkspaceName = WorkspaceName(namespace,name)
```
Comment (cahrens) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: Nit: missing space after the comma.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/ApiServiceSpec.scala [hunk 4]
```diff
@@ -196,6 +196,7 @@ trait ApiServiceSpec extends TestDriverComponentWithFlatSpecAndMatchers with Raw
       true,
       "fc-"
     )
+    val multiCloudWorkspaceConfig = MultiCloudWorkspaceConfig(testConf)
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/ApiServiceSpec.scala]: I would suggest either naming this `testMultiCloudWorkspaceConfig`, or just inlining it on line 242. As it is, I was confused why we needed to override `MultiCloudWorkspaceService.constructor`.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala [hunk 5]
```diff
@@ -342,6 +342,7 @@ class SubmissionSpec(_system: ActorSystem) extends TestKit(_system)
         trackDetailedSubmissionMetrics = true,
         "fc-"
       )
+      val multiCloudWorkspaceConfig = MultiCloudWorkspaceConfig(testConf)
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: Why is this needed here?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+  def createMultiCloudWorkspace(workspaceRequest: MultiCloudWorkspaceRequest, parentSpan: Span = null) = {
+    if (!multiCloudWorkspaceConfig.multiCloudWorkspacesEnabled) {
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotImplemented, "MC workspaces are not enabled"))
+    }
+
+    for {
+      workspace <- traceWithParent("withNewMcWorkspaceContext", parentSpan)(
+        opSpan => dataSource.inTransactionWithAttrTempTable(Set(AttributeTempTableType.Workspace))({ dataAccess =>
+          createWorkspace(workspaceRequest, dataAccess, opSpan) { createdWorkspace =>
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: Per our discussion today, does there need to be cleanup on deletion?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+      workspaceId = workspaceId,
+      bucketName = "",
+      workflowCollectionName = Some(workspaceId),
+      createdDate = currentDate,
+      lastModified = currentDate,
+      createdBy = userInfo.userEmail.value,
+      attributes = workspaceRequest.attributes,
+      isLocked = false,
+      workspaceVersion = WorkspaceVersions.V2,
+      googleProjectId = GoogleProjectId(""),
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: It's weird to see all these Google properties in a MC Workspace. Is it necessary to specify these dummy values, or can default be used (or another constructor added).

**Review by cahrens — COMMENTED** [review]
I took a first pass through this. TBH, a lot of it seems pretty magical to me. I'll take another look on Tuesday, and hopefully more will sink in then.

**Comment by cahrens** [comment]
@aherbst-broad are the test failures expected?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala (hunk 7 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: 👍 I'll investigate a custom constructor

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala (hunk 5 above)
Comment (aherbst-broad) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: It is not, will remove.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/ApiServiceSpec.scala (hunk 4 above)
Comment (aherbst-broad) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/ApiServiceSpec.scala]: 👍 refactored. The constructor is needed as the `ApiServices` trait in this spec implements `RawlsApiService`, so any new service that's added to that trait needs to be implemented here as well.

**Comment by aherbst-broad** [comment]
@cahrens Actually, it appears that the more idiomatic approach for case classes is to overload `apply` on the companion object, I've switched to that. 

More broadly, Scala does support the notion of class methods, via a [companion object](https://docs.scala-lang.org/overviews/scala-book/companion-objects.html)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala (hunk 6 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: I expect that will be the case when we implement WOR-80 (building the deletion endpoint). As of now, this endpoint isn't allocating any resources other than a workspace + cloud context record in WSM.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+
+// TODO this data will be pulled from the spend profile service, hardcoding in conf until that svc is ready
+final case class MultiCloudWorkspaceConfig(multiCloudWorkspacesEnabled: Boolean,
+                                           cloudContextPollTimeout: FiniteDuration,
+                                           azureConfig: Option[AzureConfig])
+
+final case class AzureConfig(spendProfileId: String,
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:16]: Which type of spend profile is this? ie does WSM have its own spend profiles, is this referencing an id from the new Spend Profile Service, or is this the TDR version?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiService.scala [hunk 9]
```diff
@@ -47,6 +48,22 @@ trait WorkspaceApiService extends UserInfoDirectives {
...
+    } ~
+      path("workspaces" / "mc" ) {
+        post {
+          entity(as[MultiCloudWorkspaceRequest]) { workspace =>
+            addLocationHeader(workspace.path) {
+              traceRequest { span =>
+                complete {
+                  multiCloudWorkspaceServiceConstructor(userInfo)
+                    .createMultiCloudWorkspace(workspace, span).map {
+                      w => StatusCodes.Created -> WorkspaceDetails(w, workspace.authorizationDomain.getOrElse(Set.empty))
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiService.scala]: Does WSM support auth domains in Azure workspaces? If not, I might return an error if a request comes in with an auth domain

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            Future(workspaceManagerDAO.createWorkspaceWithSpendProfile(workspaceId, workspaceRequest.name, spendProfileId, userInfo.accessToken))
+          ))
+          _ = logger.info(s"Creating cloud context in WSM [workspaceId = ${workspaceId}]")
+          cloudContextCreateResult <- traceDBIOWithParent("createCloudContext", parentSpan)(_ => DBIO.from(
+            Future(workspaceManagerDAO.createAzureWorkspaceCloudContext(workspaceId, azureTenantId, azureResourceGroupId, azureSubscriptionId, userInfo.accessToken))
+          ))
+          jobControlId = cloudContextCreateResult.getJobReport.getId
+          _ = logger.info(s"Polling on cloud context in WSM [jobControlId = ${jobControlId}]")
+          pollingResult <- traceDBIOWithParent("pollCloudContext", parentSpan)(_ => DBIO.from(
+            pollCloudContext(workspaceId, cloudContextCreateResult.getJobReport.getId, userInfo.accessToken)
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: nit: use `jobControlId` here

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+          _ <- traceDBIOWithParent("createWorkspaceInWSM", parentSpan)(_ => DBIO.from(
+            Future(workspaceManagerDAO.createWorkspaceWithSpendProfile(workspaceId, workspaceRequest.name, spendProfileId, userInfo.accessToken))
+          ))
+          _ = logger.info(s"Creating cloud context in WSM [workspaceId = ${workspaceId}]")
+          cloudContextCreateResult <- traceDBIOWithParent("createCloudContext", parentSpan)(_ => DBIO.from(
+            Future(workspaceManagerDAO.createAzureWorkspaceCloudContext(workspaceId, azureTenantId, azureResourceGroupId, azureSubscriptionId, userInfo.accessToken))
+          ))
+          jobControlId = cloudContextCreateResult.getJobReport.getId
+          _ = logger.info(s"Polling on cloud context in WSM [jobControlId = ${jobControlId}]")
+          pollingResult <- traceDBIOWithParent("pollCloudContext", parentSpan)(_ => DBIO.from(
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: nit: since we don't use `pollingResult` anywhere, change this to a `_`. Or maybe we should return the `CreateCloudContextResult` we get when polling `getCloudContextCreationStatus`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    val workspace = Workspace(
+      namespace = workspaceRequest.namespace,
+      name = workspaceRequest.name,
+      workspaceId = workspaceId,
+      createdDate = currentDate,
+      lastModified = currentDate,
+      createdBy = userInfo.userEmail.value,
+      attributes = workspaceRequest.attributes
+    )
+    traceDBIOWithParent("saveMultiCloudWorkspace", parentSpan)(_ => dataAccess.workspaceQuery.createOrUpdate(workspace))
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:152]: should this include the workspace type?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+            Future(workspaceManagerDAO.createAzureWorkspaceCloudContext(workspaceId, azureTenantId, azureResourceGroupId, azureSubscriptionId, userInfo.accessToken))
+          ))
+          jobControlId = cloudContextCreateResult.getJobReport.getId
+          _ = logger.info(s"Polling on cloud context in WSM [jobControlId = ${jobControlId}]")
+          pollingResult <- traceDBIOWithParent("pollCloudContext", parentSpan)(_ => DBIO.from(
+            pollCloudContext(workspaceId, cloudContextCreateResult.getJobReport.getId, userInfo.accessToken)
+          ))
+          _ = logger.info(s"Creating workspace record [workspaceId = ${workspaceId}]")
+          savedWorkspace <- traceDBIOWithParent("saveNewWorkspaceToRawlsDB", parentSpan)(_ => {
+            createMultiCloudWorkspaceInDatabase(
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: Can we limit the scope of the DB transaction to this call rather than including the WSM calls in the transaction?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala (hunk 8 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:16]: WSM has it's own hardcoded spend profile for now but this will eventually be sourced from the new spend profile service.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala (hunk 11 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: _ makes sense, I'll make the change

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala (hunk 12 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:152]: 🤔 Not sure I follow, the name of the span indicates that it's multicloud -- are you saying we should also interpolate `workspace.workspaceType` into the span name?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+
+case object MultiCloudWorkspaceConfig {
+  def apply[T <: MultiCloudWorkspaceConfig](conf: Config): MultiCloudWorkspaceConfig = {
+    val azureConfig: Option[AzureConfig] = conf.getConfigOption("multiCloudWorkspaces.azureConfig") match {
+      case Some(azc) => Some(AzureConfig(
+        azc.getString("spendProfileId"),
+        azc.getString("tenantId"),
+        azc.getString("subscriptionId"),
+        azc.getString("resourceGroupId")
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:29]: Looks like there is some missing test coverage?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+    conf.getConfigOption("multiCloudWorkspaces") match {
+      case Some(mc) =>
+        new MultiCloudWorkspaceConfig(
+          conf.getBoolean("multiCloudWorkspaces.enabled"),
+          util.toScalaDuration(conf.getDuration("multiCloudWorkspaces.cloudContextPollTimeoutSeconds")),
+          azureConfig
+        )
+      case None =>
+        new MultiCloudWorkspaceConfig(false, 2 seconds, None)
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:42]: The "2 seconds" here seems crazy to me (as syntax)… but I assume it works.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/workspacemanager/HttpWorkspaceManagerDAO.scala [hunk 16]
```diff
@@ -41,6 +41,35 @@ class HttpWorkspaceManagerDAO(baseWorkspaceManagerUrl: String)(implicit val syst
     getWorkspaceApi(accessToken).createWorkspace(new CreateWorkspaceRequestBody().id(workspaceId))
   }
 
+  override def createWorkspaceWithSpendProfile(workspaceId: UUID,
+                                               displayName: String,
+                                               spendProfileId: String,
+                                               accessToken: OAuth2BearerToken): CreatedWorkspace = {
+    getWorkspaceApi(accessToken).createWorkspace(new CreateWorkspaceRequestBody()
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/workspacemanager/HttpWorkspaceManagerDAO.scala:48]: More code coverage warnings on this code.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+  private def jobStatusPredicate(t: Throwable): Boolean = {
+    t match {
+      case t: WorkspaceManagerPollingOperationException => (t.status == StatusEnum.RUNNING)
+      case _ => false
+    }
+  }
+
+  private def pollCloudContext(workspaceId: UUID, jobControlId: String, accessToken: OAuth2BearerToken): Future[Unit] = {
+    for {
+      result <- retryUntilSuccessOrTimeout(pred = jobStatusPredicate)(2 seconds, multiCloudWorkspaceConfig.cloudContextPollTimeout) {
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:132]: Why is 2 seconds hardcoded here?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala (hunk 15 above)
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:42]: Perhaps the code coverage warnings are because the tests didn't run on the last commit?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala (hunk 14 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:29]: Added a small test for this class.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala (hunk 15 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:42]: We get the postfix syntax by importing `import scala.language.postfixOps`, I think this is consistent with the rest of the codebase.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+  def createMultiCloudWorkspace(workspaceRequest: MultiCloudWorkspaceRequest, parentSpan: Span = null) = {
+    if (!multiCloudWorkspaceConfig.multiCloudWorkspacesEnabled) {
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotImplemented, "MC workspaces are not enabled"))
+    }
+
+    for {
+      workspace <- traceWithParent("createMultiCloudWorkspace", parentSpan)(s1 =>
+        createWorkspace(workspaceRequest, s1))
+
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: Nit: extra line.

**Review by cahrens — APPROVED** [review]
I've looked through the code a couple of times, but feel free to have someone else do an additional review. I'm not familiar with a lot of things in this code (and I never would have thought of the good questions that Mark raised!), so having someone else with more familiarity look at the code might be good.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 19]
```diff
@@ -56,55 +55,49 @@ class MultiCloudWorkspaceService(userInfo: UserInfo,
...
-        opSpan => dataSource.inTransactionWithAttrTempTable(Set(AttributeTempTableType.Workspace))({ dataAccess =>
-          createWorkspace(workspaceRequest, dataAccess, opSpan) { createdWorkspace =>
-            DBIO.successful(createdWorkspace)
-          }
-        }, TransactionIsolation.ReadCommitted)) // read committed to avoid deadlocks on workspace attr scratch table
-    } yield workspace
+      workspace <- traceWithParent("createMultiCloudWorkspace", parentSpan)(s1 =>
+        createWorkspace(workspaceRequest, s1))
 
+    } yield workspace
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala]: You could simplify this by removing the for comprehension entirely and just having 

`traceWithParent("createMultiCloudWorkspace", parentSpan)(s1 =>
        createWorkspace(workspaceRequest, s1))`

**Review by marctalbott — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    conf.getConfigOption("multiCloudWorkspaces") match {
+      case Some(mc) =>
+        new MultiCloudWorkspaceConfig(
+          conf.getBoolean("multiCloudWorkspaces.enabled"),
+          util.toScalaDuration(conf.getDuration("multiCloudWorkspaces.cloudContextPollTimeoutSeconds")),
+          azureConfig
+        )
+      case None =>
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/config/MultiCloudWorkspaceConfig.scala:41]: another option (which is more typical in rawls) is to add these defaults to `rawls/core/src/main/resources/reference.conf`

**Review by MatthewBemis — APPROVED** [review]
This all looks reasonable to me 👍
